components/container_id/container_id.c

```diff
@@ -1,101 +1,113 @@
-        --*end;
-        --len;
-    }
-    return len;
-}
+#define CONTAINER_REGEX "[0-9a-f]{64}"
+/* Example of a valid task ID.
+ *
+ * Fargate 1.4+
+ * 1:name=systemd:/ecs/34dc0b5e626f2c5c4c5170e34b10e765-1234567890
+ */
+#define TASK_REGEX "[0-9a-f]{32}-[0-9]+"
 
-static void dd_extract_id(char *buf, char *pos) {
-    char *end = NULL;
-    size_t len = dd_str_seek_end(pos, &end);
+#define MIN_ID_LEN DATADOG_PHP_CONTAINER_ID_MIN_LEN
+#define MAX_ID_LEN DATADOG_PHP_CONTAINER_ID_MAX_LEN
 
-    if (len < DATADOG_PHP_CONTAINER_ID_MAX_LEN) {
-        return;
-    }
+static size_t dd_find_id_regex(regex_t *regex, char *buf, const char *line) {
+    size_t len;
+    regmatch_t pmatch[1];
 
-    // TODO Refactor to support Fargate 1.4+ IDs
-    // [a-f0-9]{64}
-    for (size_t i = DATADOG_PHP_CONTAINER_ID_MAX_LEN; i > 0; i--) {
-        char c = end[0];
-        if (!isxdigit(c)) {
-            buf[0] = '\0';
-            return;
-        }
-        buf[i - 1] = c;
-        end--;
-    }
-    buf[DATADOG_PHP_CONTAINER_ID_MAX_LEN] = '\0';
-}
+    if (regexec(regex, line, 1, pmatch, 0) != 0) return 0;
 
-static char *dd_find_target(const char *line) {
-    for (size_t i = 0; i < dd_targets_len; ++i) {
-        char *pos = strstr(line, dd_targets[i].name);
-        if (pos) {
-            return pos + dd_targets[i].len;
-        }
-    }
-    return NULL;
+    len = (size_t)pmatch[0].rm_eo - pmatch[0].rm_so;
+    if (len < MIN_ID_LEN || len > MAX_ID_LEN) return 0;
+
+    memcpy(buf, line + pmatch[0].rm_so, len);
+    buf[len] = '\0';
+
+    return len;
 }
 
 void datadog_php_container_id(char *buf, const char *file) {
     FILE *fp;
-    char line[1024];
+    char task_buf[MAX_ID_LEN + 1];
+
+    if (buf == NULL) return;
 
     buf[0] = '\0';
+    task_buf[0] = '\0';
 
-    if (file == NULL || file[0] == '\0') {
-        return;
-    }
+    if (file == NULL || file[0] == '\0' || (fp = fopen(file, "r")) == NULL) return;
 
-    fp = fopen(file, "r");
-    if (fp == NULL) {
+    /* According to the in-code docs:
+     *
+     * "Note that the translate table must either have been initialized by
+     * 'regcomp', with a malloc'ed value, or set to NULL before calling
+     * 'regfree'."
+     * https://elixir.bootlin.com/glibc/latest/source/posix/regex.h#L535
+     *
+     * For this reason we zero-initialize all of the 'regex_t' patterns so that
+     * we can still call regfree() on them if regcomp() fails to initialize the
+     * pattern buffer.
+     */
+    regex_t line_regex = {0};
+    regex_t task_regex = {0};
+    regex_t container_regex = {0};
+    int l_res = regcomp(&line_regex, LINE_REGEX, REG_EXTENDED | REG_NOSUB);
+    int t_res = regcomp(&task_regex, TASK_REGEX, REG_EXTENDED);
+    int c_res = regcomp(&container_regex, CONTAINER_REGEX, REG_EXTENDED);
+    if (l_res != 0 || t_res != 0 || c_res != 0) {
+        regfree(&container_regex);
+        regfree(&task_regex);
+        regfree(&line_regex);
+        fclose(fp);
```

Comment:
I'm not going to block this PR, but whenever there is more than 1 thing to cleanup on an early return I try to re-examine the structure to try to make it cleaner (if possible without goto). Can you put some time on my calendar to go over this?

Reply:
The funny thing is, this is the cleaned up version; you should have seen it before. 😆 But absolutely, if you have some ideas on how to make this even cleaner, I'd love to hear your thoughts.